js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,479 @@
+	if lib.GetState(*ctxPtr) != nil {
+		return nil, errors.New("load must be called in the init context")
+	}
+
+	parser := protoparse.Parser{
+		ImportPaths:      importPaths,
+		InferImportPaths: len(importPaths) == 0,
+	}
+
+	fds, err := parser.ParseFiles(filenames...)
+	if err != nil {
+		return nil, err
+	}
+
+	var rtn []MethodInfo
+	c.mds = make(map[string]*desc.MethodDescriptor)
+	for _, fd := range fds {
+		for _, sd := range fd.GetServices() {
+			for _, md := range sd.GetMethods() {
+				var s strings.Builder
+				s.WriteString(sd.GetFullyQualifiedName())
+				s.WriteRune('/')
+				s.WriteString(md.GetName())
+				name := s.String()
+				c.mds[name] = md
+				rtn = append(rtn, MethodInfo{
+					FullMethod:     name,
+					IsClientStream: md.IsClientStreaming(),
+					IsServerStream: md.IsServerStreaming(),
+				})
+			}
+		}
+	}
+
+	return rtn, nil
+}
+
+type transportCreds struct {
+	credentials.TransportCredentials
+	errc chan<- error
+}
+
+func (t transportCreds) ClientHandshake(ctx context.Context,
+	addr string, in net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	out, auth, err := t.TransportCredentials.ClientHandshake(ctx, addr, in)
+	if err != nil {
+		t.errc <- err
+	}
+
+	return out, auth, err
+}
+
+// Connect is a block dial to the gRPC server at the given address (host:port)
+func (c *Client) Connect(ctxPtr *context.Context, addr string, params map[string]interface{}) (bool, error) {
+	state := lib.GetState(*ctxPtr)
+	if state == nil {
+		return false, errConnectInInitContext
+	}
+
+	isPlaintext, timeout := false, 60*time.Second
+
+	for k, v := range params {
+		switch k {
+		case "plaintext":
+			isPlaintext, _ = v.(bool)
+		case "timeout":
+			if t, ok := v.(float64); ok && t > 0.0 {
+				timeout = time.Duration(t) * time.Millisecond
+			}
+		}
+	}
+
+	errc := make(chan error)
+	go func() {
+		opts := []grpc.DialOption{
+			grpc.WithBlock(),
+			grpc.FailOnNonTempDialError(true),
+			grpc.WithStatsHandler(c),
+		}
+
+		if ua := state.Options.UserAgent; ua.Valid {
+			opts = append(opts, grpc.WithUserAgent(ua.ValueOrZero()))
+		}
+
+		if !isPlaintext {
+			tlsCfg := state.TLSConfig.Clone()
+			tlsCfg.NextProtos = []string{"h2"}
+
+			// TODO(rogchap): Would be good to add support for custom RootCAs (self signed)
+
+			// (rogchap) we create a wrapper for transport credentials so that we can report
+			// on any TLS errors.
+			creds := transportCreds{
+				credentials.NewTLS(tlsCfg),
+				errc,
+			}
+			opts = append(opts, grpc.WithTransportCredentials(creds))
+		}
+
+		if isPlaintext {
+			opts = append(opts, grpc.WithInsecure())
+		}
+
+		dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+			return state.Dialer.DialContext(ctx, "tcp", addr)
+		}
+		opts = append(opts, grpc.WithContextDialer(dialer))
+
+		ctx, cancel := context.WithTimeout(*ctxPtr, timeout)
+		defer cancel()
+
+		var err error
+		c.conn, err = grpc.DialContext(ctx, addr, opts...)
+		if err != nil {
+			errc <- err
```

Comment:
I think you are right this will leak the goroutine. This has not been a problem so far, because the CLI will always exit after an error like this; but would be an issue if running as a service. I think the best thing to do is create a buffered channel so that it does not block and then the channel will get garbage collected as soon as this goroutine returns.